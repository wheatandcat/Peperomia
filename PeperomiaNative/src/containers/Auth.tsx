import * as GoogleSignIn from "expo-google-sign-in";
import * as Google from "expo-google-app-auth";
import { AsyncStorage, Platform } from "react-native";
import React, { createContext, Component } from "react";
import Constants from "expo-constants";
import { Sentry } from "react-native-sentry";
import * as firebase from "firebase";

const Context = createContext({});
const { Provider } = Context;

interface Props {}

interface State {
  uid: string;
  email: string;
}

const isStandaloneAndAndroid = () => {
  return Platform.OS === "android" && Constants.appOwnership !== "expo";
};

export default class extends Component<Props, State> {
  state = {
    email: "",
    uid: ""
  };

  async componentDidMount() {
    if (isStandaloneAndAndroid()) {
      const androidClientId = process.env.GOOGLE_LOGIN_ANDROID_CLIENT_ID;
      try {
        await GoogleSignIn.initAsync({
          clientId: String(androidClientId)
        });
      } catch ({ message }) {
        Sentry.captureMessage(JSON.stringify(message));
      }
    }

    const loggedIn = await this.loggedIn();

    if (loggedIn && !this.state.uid) {
      const uid = await AsyncStorage.getItem("uid");
      if (uid) {
        this.setState({
          uid
        });
      }
    }

    if (loggedIn && !this.state.email) {
      const email = await AsyncStorage.getItem("email");
      if (email) {
        this.setState({
          email
        });
      }
    }
  }

  onGoogleLogin = async () => {
    if (isStandaloneAndAndroid()) {
      // TODO: AndroidのstandaloneのみGoogleSignInを使わないとエラーになる
      // https://github.com/expo/expo/issues/4762

      await GoogleSignIn.askForPlayServicesAsync();
      const result = await GoogleSignIn.signInAsync();

      if (result.type === "success" && result.user && result.user.auth) {
        const { idToken, accessToken } = result.user.auth;
        await this.firebaseLogin(idToken || "", accessToken || "");
      } else {
        Sentry.captureMessage(JSON.stringify(result));
      }
    } else {
      const androidClientId = process.env.GOOGLE_LOGIN_ANDROID_CLIENT_ID;
      const iosClientId = process.env.GOOGLE_LOGIN_IOS_CLIENT_ID;
      const result = await Google.logInAsync({
        clientId:
          Platform.OS === "ios" ? String(iosClientId) : String(androidClientId),
        iosClientId,
        androidClientId,
        scopes: ["profile", "email"]
      });

      if (result.type === "success") {
        const { idToken, accessToken } = result;
        await this.firebaseLogin(idToken || "", accessToken || "");
      } else {
        Sentry.captureMessage(JSON.stringify(result));
      }
    }
  };

  firebaseLogin = async (idToken: string, accessToken: string) => {
    const credential = firebase.auth.GoogleAuthProvider.credential(
      idToken,
      accessToken
    );

    await firebase
      .auth()
      .signInAndRetrieveDataWithCredential(credential)
      .catch(error => {
        Sentry.captureMessage(JSON.stringify(error));
      });

    await this.setSession(true);
  };

  loggedIn = async () => {
    const idToken = await this.getIdToken();

    return Boolean(idToken);
  };

  logout = async () => {
    await firebase.auth().signOut();

    await AsyncStorage.removeItem("id_token");
    await AsyncStorage.removeItem("expiration");
    await AsyncStorage.removeItem("email");
  };

  setSession = async (refresh = false) => {
    const user = firebase.auth().currentUser;
    if (!user) {
      return;
    }

    if (user.email) {
      await AsyncStorage.setItem("email", user.email);
      await AsyncStorage.setItem("uid", user.uid);
      this.setState({
        email: user.email,
        uid: user.uid
      });
    }

    const idToken = await user.getIdToken(refresh);
    await AsyncStorage.setItem("id_token", idToken);
    await AsyncStorage.setItem(
      "expiration",
      String(new Date().getTime() + 60 * 60)
    );

    return idToken;
  };

  getIdToken = async () => {
    const idToken = await AsyncStorage.getItem("id_token");
    if (!idToken) {
      return null;
    }

    const expiration = await AsyncStorage.getItem("expiration");
    if (Number(expiration) > new Date().getTime()) {
      return idToken;
    }

    return this.setSession(true);
  };

  render() {
    return (
      <Provider
        value={{
          onGoogleLogin: this.onGoogleLogin,
          getIdToken: this.getIdToken,
          loggedIn: this.loggedIn,
          logout: this.logout,
          email: this.state.email,
          uid: this.state.uid
        }}
      >
        {this.props.children}
      </Provider>
    );
  }
}

export const Consumer = Context.Consumer;
