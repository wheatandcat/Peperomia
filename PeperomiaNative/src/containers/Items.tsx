import { SQLite } from "expo-sqlite";
import React, { createContext, Component } from "react";
import { db } from "../lib/db";
import { select as selectItems, Item } from "../lib/db/item";
import {
  selectByItemId as selectItemDetailByItemId,
  ItemDetail
} from "../lib/db/itemDetail";

const Context = createContext({});
const { Provider } = Context;

interface ItemAbout {
  itemId: number;
  about: string;
}

interface Props {}

interface State {
  loading: boolean;
  items: Item[];
  about: ItemAbout[];
}

export default class extends Component<Props, State> {
  state = {
    loading: true,
    items: [],
    about: []
  };

  componentDidMount() {
    this.getData();
  }

  getData = () => {
    db.transaction((tx: SQLite.Transaction) => {
      selectItems(tx, this.setItems);
    });
  };

  setItems = (data: any, error: any) => {
    if (error) {
      return;
    }
    this.setState({
      items: data,
      about: []
    });

    data.map((val: Item) => {
      db.transaction((tx: SQLite.Transaction) => {
        selectItemDetailByItemId(tx, String(val.id), this.setItemsDetail);
      });
    });
  };

  setItemsDetail = (data: any, error: any) => {
    if (error || !data || data.length === 0) {
      return;
    }

    const names = data.map((val: ItemDetail) => val.title).join("â†’");
    const itemId = data[0].itemId;
    const about = [
      ...this.state.about,
      {
        itemId: itemId,
        about: names
      }
    ];

    this.setState({
      about,
      loading: false
    });
  };

  render() {
    return (
      <Provider
        value={{
          items: this.state.items,
          about: this.state.about,
          refreshData: this.getData,
          itemsLoading: this.state.loading
        }}
      >
        {this.props.children}
      </Provider>
    );
  }
}

export const Consumer = Context.Consumer;
