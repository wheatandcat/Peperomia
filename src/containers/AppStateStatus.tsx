import React, { useCallback, useEffect, useState } from 'react';
import { AppState, AppStateStatus } from 'react-native';

type Props = {};

type State = {
  loading: boolean;
};

const AppStateContainer: React.FC<Props> = (props) => {
  const [appState, setAppState] = useState<AppStateStatus>(
    AppState.currentState
  );

  const check = useCallback(async () => {}, []);

  const handleAppStateChange = useCallback(
    async (nextAppState) => {
      if (appState.match(/inactive|background/u) && nextAppState === 'active')
        await check();

      setAppState(nextAppState);
    },
    [appState, check]
  );

  useEffect(() => {
    AppState.addEventListener('change', handleAppStateChange);

    return () => {
      AppState.removeEventListener('change', handleAppStateChange);
    };
  }, [handleAppStateChange]);

  return <>{props.children}</>;
};

export default AppStateContainer;
