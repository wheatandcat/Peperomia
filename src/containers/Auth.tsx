import * as GoogleSignIn from 'expo-google-sign-in';
import * as Google from 'expo-google-app-auth';
import { Platform, Alert } from 'react-native';
import AsyncStorage from '@react-native-community/async-storage';
import uuidv4 from 'uuid/v4';
import dayjs from 'dayjs';
import advancedFormat from 'dayjs/plugin/advancedFormat';
import 'dayjs/locale/ja';
import * as AppleAuthentication from 'expo-apple-authentication';
import React, {
  memo,
  createContext,
  FC,
  useState,
  useCallback,
  useContext,
  useEffect,
} from 'react';
import { View } from 'react-native';
import EStyleSheet from 'react-native-extended-stylesheet';
import Spinner from 'react-native-loading-spinner-overlay';
import Constants from 'expo-constants';
import { backup } from 'lib/backup';
import * as Sentry from 'sentry-expo';
import firebase from 'lib/system/firebase';
import useIsFirstRender from 'hooks/useIsFirstRender';
import { UID } from 'domain/user';
import libAuth from 'lib/auth';
import {
  useSyncCalendarsMutation,
  SyncCalendarsMutationVariables,
  SyncCalendar,
  SyncItemDetail,
} from 'queries/api/index';
import { post } from 'lib/fetch';
import theme from 'config/theme';

dayjs.extend(advancedFormat);

const auth = new libAuth();

export const Context = createContext<ContextProps>({});
const { Provider } = Context;

type Props = {};

type State = {
  uid: UID;
  setup: boolean;
  render: boolean;
  email: string;
};

export type ContextProps = Partial<
  Pick<State, 'email' | 'uid'> & {
    onGoogleLogin: () => Promise<string | null>;
    onAppleLogin: () => Promise<string | null>;
    getIdToken: () => Promise<string | null>;
    loggedIn: () => Promise<boolean>;
    logout: () => Promise<void>;
  }
>;

const isStandaloneAndAndroid = () => {
  return Platform.OS === 'android' && Constants.appOwnership !== 'expo';
};

const Auth: FC<Props> = memo((props) => {
  const [state, setState] = useState<State>({
    email: '',
    uid: null,
    render: true,
    setup: false,
  });
  const isFirstRender = useIsFirstRender();
  const [syncCalendarsMutation] = useSyncCalendarsMutation({
    async onCompleted() {
      setRender();
    },
    onError(err) {
      Alert.alert('バックアップに失敗しました', err.message);
    },
  });

  const getIdToken = useCallback(async () => {
    return await auth.getIdToken();
  }, []);

  const setRender = useCallback(() => {
    setState((s) => ({
      ...s,
      render: true,
    }));
  }, []);

  const saveUser = useCallback(async () => {
    const idToken = await getIdToken();

    const response = await post('CreateUser', null, idToken || '');

    if (response?.error) {
      Alert.alert('ユーザーの保存に失敗しました。');
    }

    return Number(response?.status);
  }, [getIdToken]);

  const backupData = useCallback(async () => {
    const httpStatus = await saveUser();

    if (httpStatus === 201) {
      // ユーザーを新規で作成した場合はデータをサーバーに送る
      const { items, itemDetails, calendars } = await backup();

      const syncCalendars = calendars.map((v) => {
        const i = items.find((v1) => v1.id === v.itemId);
        let ids: SyncItemDetail[] = [];

        if (i) {
          ids = itemDetails
            .filter((v2) => v2.itemId === v.itemId)
            .map((id) => ({
              id: uuidv4(),
              title: id.title || '',
              kind: id?.kind || '',
              place: id?.place || '',
              url: id?.url || '',
              memo: id?.memo || '',
              priority: id?.priority || 1,
            }));
        }

        const r: SyncCalendar = {
          id: uuidv4(),
          date: dayjs(v.date).format('YYYY-MM-DDT00:00:00'),
          item: {
            id: uuidv4(),
            title: i?.title || '',
            kind: i?.kind || '',
            itemDetails: ids,
          },
        };

        return r;
      });

      const variables: SyncCalendarsMutationVariables = {
        calendars: {
          calendars: syncCalendars,
        },
      };

      syncCalendarsMutation({ variables });
    } else if (httpStatus === 200) {
      setRender();
    } else {
      Alert.alert('エラー', 'ユーザー登録/ログインに失敗しました', [
        {
          text: '閉じる',
          onPress: () => {
            setRender();
          },
        },
      ]);
    }
  }, [saveUser, syncCalendarsMutation, setRender]);

  const setSession = useCallback(
    async (refresh = false) => {
      const idToken = await auth.setSession(refresh);

      if (idToken) {
        const user = firebase.auth().currentUser;
        if (user) {
          setState((s) => ({
            ...s,
            render: false,
          }));
          await AsyncStorage.setItem('email', user?.email || '');
          await AsyncStorage.setItem('uid', user.uid);
          setState((s) => ({
            ...s,
            email: user.email || '',
            uid: user.uid || '',
          }));

          await backupData();
        }
      }

      return idToken;
    },
    [backupData]
  );

  const loggedIn = useCallback(async () => {
    const idToken = await getIdToken();

    return Boolean(idToken);
  }, [getIdToken]);

  const firebaseLogin = useCallback(
    async (credential: firebase.auth.OAuthCredential) => {
      const data = await firebase
        .auth()
        .signInWithCredential(credential)
        .catch((error: any) => {
          console.log(error);
        });
      console.log(data);

      return await setSession(true);
    },
    [setSession]
  );

  const firebaseGoogleLogin = useCallback(
    async (idToken: string, accessToken: string) => {
      const credential = firebase.auth.GoogleAuthProvider.credential(
        idToken,
        accessToken
      );

      await firebaseLogin(credential);
    },
    [firebaseLogin]
  );

  const onAppleLogin = useCallback(async () => {
    const nonceString = nonceGen(32);

    try {
      const result = await AppleAuthentication.signInAsync({
        requestedScopes: [
          AppleAuthentication.AppleAuthenticationScope.FULL_NAME,
          AppleAuthentication.AppleAuthenticationScope.EMAIL,
        ],
        state: nonceString,
      });
      const provider = new firebase.auth.OAuthProvider('apple.com');
      const credential = provider.credential({
        idToken: result.identityToken || '',
        rawNonce: nonceString,
      });

      await firebaseLogin(credential);
      return await AsyncStorage.getItem('uid');
    } catch (e) {
      console.log(e);
      return null;
    }
  }, [firebaseLogin]);

  const onGoogleLogin = useCallback(async () => {
    if (isStandaloneAndAndroid()) {
      // TODO: AndroidのstandaloneのみGoogleSignInを使わないとエラーになる
      // https://github.com/expo/expo/issues/4762
      await GoogleSignIn.askForPlayServicesAsync();
      const result = await GoogleSignIn.signInAsync();

      if (result.type === 'success' && result.user && result.user.auth) {
        const { idToken, accessToken } = result.user.auth;
        await firebaseGoogleLogin(idToken || '', accessToken || '');
        return await AsyncStorage.getItem('uid');
      } else {
        Sentry.captureMessage(JSON.stringify(result));
        return null;
      }
    } else {
      const androidClientId = process.env.GOOGLE_LOGIN_ANDROID_CLIENT_ID;
      const iosClientId = process.env.GOOGLE_LOGIN_IOS_CLIENT_ID;
      const result = await Google.logInAsync({
        clientId:
          Platform.OS === 'ios' ? String(iosClientId) : String(androidClientId),
        iosClientId,
        androidClientId,
        scopes: ['profile', 'email'],
      });

      console.log(result);

      if (result.type === 'success') {
        const { idToken, accessToken } = result;
        await firebaseGoogleLogin(idToken || '', accessToken || '');
        return await AsyncStorage.getItem('uid');
      } else {
        Sentry.captureMessage(JSON.stringify(result));
        return null;
      }
    }
  }, [firebaseGoogleLogin]);

  useEffect(() => {
    if (!isFirstRender) return;

    if (isStandaloneAndAndroid()) {
      const androidClientId = process.env.GOOGLE_LOGIN_ANDROID_CLIENT_ID;
      try {
        GoogleSignIn.initAsync({
          clientId: String(androidClientId),
        });
      } catch ({ message }) {
        Sentry.captureMessage(JSON.stringify(message));
      }
    }

    const checkLogin = async () => {
      const login = await loggedIn();

      if (login && !state.uid) {
        const uid = await AsyncStorage.getItem('uid');

        if (uid) {
          setState((s) => ({
            ...s,
            uid,
            setup: true,
          }));
        }
      } else {
        setState((s) => ({
          ...s,
          setup: true,
        }));
      }

      if (login && !state.email) {
        const email = await AsyncStorage.getItem('email');
        if (email) {
          setState((s) => ({
            ...s,
            email,
          }));
        }
      }
    };

    firebase.auth().onAuthStateChanged(() => {
      checkLogin();
    });
  }, [isFirstRender, state, loggedIn]);

  const onLogout = useCallback(async () => {
    setState((s) => ({
      ...s,
      render: false,
    }));

    await logout();
    setState((s) => ({
      ...s,
      email: '',
      uid: null,
      render: true,
    }));
  }, []);

  if (!state.setup) {
    return null;
  }

  if (!state.render) {
    return (
      <View style={styles.container}>
        <Spinner
          visible={true}
          textContent="読込中"
          textStyle={{ color: theme().color.white }}
        />
      </View>
    );
  }

  return (
    <Provider
      value={{
        onGoogleLogin: onGoogleLogin,
        onAppleLogin: onAppleLogin,
        getIdToken: getIdToken,
        loggedIn: loggedIn,
        logout: onLogout,
        email: state.email,
        uid: state.uid,
      }}
    >
      {props.children}
    </Provider>
  );
});

const logout = async () => {
  await firebase.auth().signOut();

  await AsyncStorage.removeItem('id_token');
  await AsyncStorage.removeItem('expiration');
  await AsyncStorage.removeItem('email');
};

export const Consumer = Context.Consumer;
export const useAuth = () => useContext(Context);

export default Auth;

const nonceGen = (length: number) => {
  let result = '';
  let characters =
    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
  let charactersLength = characters.length;
  for (let i = 0; i < length; i++) {
    result += characters.charAt(Math.floor(Math.random() * charactersLength));
  }
  return result;
};

const styles = EStyleSheet.create({
  container: {
    alignItems: 'center',
    backgroundColor: '$background',
    height: '100%',
  },
});
