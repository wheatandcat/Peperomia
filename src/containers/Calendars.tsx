import React, { createContext, useState, useCallback, useContext } from 'react';
import dayjs from 'dayjs';
import advancedFormat from 'dayjs/plugin/advancedFormat';
import 'dayjs/locale/ja';
import { CalendarsQueryResult, CalendarsQuery } from 'queries/api/index';
import useCalendarsHooks from 'hooks/useCalendars';

dayjs.extend(advancedFormat);

export const Context = createContext<ContextProps>({
  loadingCalendars: false,
  calendars: [],
  refetchCalendars: undefined,
  errorCalendars: undefined,
  setDate: (_1: string, _2: string) => null,
});
const { Provider } = Context;

export type Calendars = NonNullable<CalendarsQuery['calendars']>;

type Props = {};

type State = {
  startDate: string;
  endDate: string;
};

const initialState = () => {
  return {
    startDate: dayjs().startOf('month').format('YYYY-MM-DDT00:00:00'),
    endDate: dayjs().endOf('month').format('YYYY-MM-DDT23:59:59'),
  };
};

export type ContextProps = {
  loadingCalendars: boolean;
  calendars: Calendars;
  refetchCalendars: CalendarsQueryResult['refetch'] | undefined;
  errorCalendars: CalendarsQueryResult['error'];
  setDate: (startDate: string, endDate: string) => void;
};

const Calendar: React.FC<Props> = (props) => {
  const [state, setState] = useState<State>(initialState());

  const { data, loading, error, refetch } = useCalendarsHooks({
    startDate: state.startDate,
    endDate: state.endDate,
  });

  const setDate = useCallback((startDate: string, endDate: string) => {
    setState((s) => ({
      ...s,
      startDate,
      endDate,
    }));
  }, []);

  return (
    <Provider
      value={{
        loadingCalendars: loading,
        refetchCalendars: refetch,
        calendars: data?.calendars || [],
        errorCalendars: error,
        setDate,
      }}
    >
      {props.children}
    </Provider>
  );
};

export default Calendar;

export const useCalendars = () => useContext(Context);
export const Consumer = Context.Consumer;
