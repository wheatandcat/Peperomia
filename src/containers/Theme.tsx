import React, {
  createContext,
  useState,
  useCallback,
  FC,
  useContext,
  useEffect,
} from 'react';
import { StatusBar, StyleSheet } from 'react-native';
import {
  useColorScheme,
  ColorSchemeName,
  Appearance,
} from 'react-native-appearance';
import { ThemeProvider } from 'react-native-elements';
import { setMode } from 'config/theme';
import EStyleSheet from 'react-native-extended-stylesheet';
import theme from 'config/theme';

const styles = StyleSheet.create({
  buttonStyle: {
    borderRadius: 25,
    borderColor: theme().color.secondary.main,
  },
  buttonTitleStyle: {
    color: theme().color.base.pale,
    fontWeight: '600',
  },
  dividerStyle: {
    backgroundColor: theme().color.base.light,
  },
});

const estyles = EStyleSheet.create({
  text: {
    color: '$text',
  },
  listItem: {
    backgroundColor: '$settingMenu',
  },
});

const RNETheme = {
  Button: {
    raised: true,
    buttonStyle: styles.buttonStyle,
    titleStyle: styles.buttonTitleStyle,
  },
  ListItem: {
    containerStyle: estyles.listItem,
  },
  ListItemTitle: {
    style: estyles.text,
  },
  Divider: {
    style: styles.dividerStyle,
  },
  colors: {
    primary: theme().color.primary.main,
  },
};

export const Context = createContext<ContextProps>({});
const { Provider } = Context;

type Props = {};

type State = {
  rendering: boolean;
  render: boolean;
};

export type ContextProps = Partial<
  Pick<State, 'rendering'> & {
    mode: string;
    colorScheme: string;
    onModeChange: (mode: 'light' | 'dark') => void;
    onFinishRendering: () => void;
  }
>;

const Theme: FC<Props> = (props) => {
  const scheme = useColorScheme();
  const [state, setState] = useState<State>({
    rendering: false,
    render: false,
  });

  const setUnRender = useCallback(() => {
    setState((s) => ({
      ...s,
      rendering: true,
      render: false,
    }));
  }, []);

  const setRender = useCallback(() => {
    setState((s) => ({
      ...s,
      rendering: true,
      render: true,
    }));
  }, []);

  const onModeChange = useCallback(
    async (mode: ColorSchemeName) => {
      setState((s) => ({
        ...s,
        rendering: true,
      }));

      if (mode === 'light' || mode === 'dark') {
        setMode(mode);
        setTimeout(setUnRender, 100);
        setTimeout(setRender, 101);
      }
    },
    [setRender, setUnRender]
  );

  const onFinishRendering = useCallback(() => {
    setState((s) => ({
      ...s,
      rendering: false,
    }));
  }, []);

  useEffect(() => {
    const checkMode = async () => {
      setState((s) => ({
        ...s,
        rendering: s.rendering,
        render: true,
      }));

      setMode(scheme);
    };

    const subscription = Appearance.addChangeListener(({ colorScheme }) => {
      onModeChange(colorScheme);
    });

    checkMode();

    return () => subscription.remove();
  }, [scheme, onModeChange]);

  return (
    <>
      <StatusBar backgroundColor="#fff" barStyle="light-content" />

      {state.render && (
        <ThemeProvider theme={RNETheme} useDark={scheme === 'dark'}>
          <Provider
            value={{
              colorScheme: scheme,
              rendering: state.rendering,
              mode: scheme,
              onModeChange: onModeChange,
              onFinishRendering: onFinishRendering,
            }}
          >
            {props.children}
          </Provider>
        </ThemeProvider>
      )}
    </>
  );
};

export const Consumer = Context.Consumer;
export const useTheme = () => useContext(Context);
export default Theme;
