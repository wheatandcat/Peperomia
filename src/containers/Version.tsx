import React, {
  useState,
  useEffect,
  createContext,
  useCallback,
  useContext,
} from 'react';
import AsyncStorage from '@react-native-community/async-storage';
import compareVersions from 'compare-versions';
import { migrationV104, migrationV201 } from 'lib/migration';
import { getFireStore } from 'lib/firebase';
import { getSupportVersion } from 'lib/firestore/supportVersion';
import ForceUpdate from 'components/pages/ForceUpdate/Page';

export const Context = createContext<ContextProps>({});
const { Provider } = Context;

export type ContextProps = Partial<{
  onCheckForceUpdate: () => Promise<void>;
}>;

type Props = {};

type State = {
  forceVersionUpdate: boolean;
};

const Version: React.FC<Props> = (props) => {
  const [state, setState] = useState<State>({
    forceVersionUpdate: false,
  });

  useEffect(() => {
    const checkVersion = async () => {
      let appVersion = await AsyncStorage.getItem('APP_VERSION');
      if (!appVersion) {
        appVersion = '1.0.0';
      }

      if (compareVersions.compare('1.0.5', appVersion, '>')) {
        // カラム追加のマイグレーション実行
        const ok = await migrationV104();
        if (ok) {
          // 現在のバージョンを設定
          await AsyncStorage.setItem('APP_VERSION', '1.0.5');
          appVersion = '1.0.5';
        }
      }

      if (compareVersions.compare('2.0.1', appVersion, '>')) {
        // カラム追加のマイグレーション実行
        const ok = await migrationV201();
        if (ok) {
          // 現在のバージョンを設定
          await AsyncStorage.setItem('APP_VERSION', '2.0.1');
          appVersion = '2.0.1';
        }
      }
    };

    checkVersion();
    onCheckForceUpdate();
  });

  const onCheckForceUpdate = useCallback(async () => {
    let appVersion = await AsyncStorage.getItem('APP_VERSION');
    if (!appVersion) {
      appVersion = '1.0.0';
    }

    const firestore = getFireStore();
    const supportVersion = await getSupportVersion(firestore);
    console.log(supportVersion);

    if (compareVersions.compare(supportVersion, appVersion, '>')) {
      setState((s) => ({
        ...s,
        forceVersionUpdate: true,
      }));
    }
  }, []);

  if (state.forceVersionUpdate) {
    return <ForceUpdate />;
  }

  return <Provider value={{ onCheckForceUpdate }}>{props.children}</Provider>;
};

export default Version;

export const useVersion = () => useContext(Context);
export const Consumer = Context.Consumer;
