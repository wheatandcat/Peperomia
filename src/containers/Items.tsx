import React, {
  FC,
  createContext,
  memo,
  useState,
  useCallback,
  useContext,
} from 'react';
import { SelectItem } from 'domain/item';
import { SelectCalendar } from 'domain/calendar';
import { SelectItemDetail } from 'domain/itemDetail';
import { getItems } from 'lib/item';
import { getItemDetails } from 'lib/itemDetail';
import { getCalendars } from 'lib/calendar';
import { useDidMount } from 'hooks/index';
import { UID } from 'domain/user';
import { Context as AuthContext } from './Auth';

export const Context = createContext<ContextProps>({});
const { Provider } = Context;

type ItemAbout = {
  itemId: number | string;
  about: string;
};

type Props = {};

type State = {
  loading: boolean;
  items: SelectItem[];
  itemDetails: SelectItemDetail[];
  about: ItemAbout[];
  calendars: SelectCalendar[];
};

export type ContextProps = Partial<
  Pick<State, 'items' | 'itemDetails' | 'about' | 'calendars'> & {
    refreshData: (targetUid?: UID | false) => void;
    itemsLoading: boolean;
  }
>;

const initialState = () => ({
  loading: true,
  calendarsLoading: true,
  items: [],
  itemDetails: [],
  about: [],
  calendars: [],
});

const Connected: FC<Props> = memo((props) => {
  const [state, setState] = useState<State>(initialState());
  const { uid } = useContext(AuthContext);

  const setItemsDetail = useCallback(
    (data: SelectItemDetail[]) => {
      const names = data.map((val: SelectItemDetail) => val.title).join(' â†’ ');
      const itemId = data[0].itemId;

      const about = [
        ...state.about,
        {
          itemId,
          about: names,
        },
      ];

      setState((s) => ({
        ...s,
        about,
        loading: false,
        itemDetails: [...state.itemDetails, ...data],
      }));
    },
    [state]
  );

  const getData = useCallback(
    async (targetUid: UID | false = false) => {
      setState((s) => ({
        ...s,
        loading: true,
      }));

      let uid2 = targetUid === false ? uid : targetUid;

      const items = await getItems(uid2);

      setState((s) => ({
        ...s,
        items: items,
        about: [],
      }));

      items.map(async (val) => {
        const itemDetails = await getItemDetails(uid2, String(val.id));

        if (!itemDetails || itemDetails.length === 0) {
          return;
        }

        setItemsDetail(itemDetails);
      });

      const calendars = await getCalendars(uid2);

      setState((s) => ({
        ...s,
        calendars,
        loading: false,
      }));
    },
    [setItemsDetail, uid]
  );

  useDidMount(() => {
    getData();
  });

  return (
    <Provider
      value={{
        items: state.items,
        itemDetails: state.itemDetails,
        calendars: state.calendars,
        about: state.about,
        refreshData: getData,
        itemsLoading: state.loading,
      }}
    >
      {props.children}
    </Provider>
  );
});

export default Connected;

export const useItems = () => useContext(Context);
export const Consumer = Context.Consumer;
