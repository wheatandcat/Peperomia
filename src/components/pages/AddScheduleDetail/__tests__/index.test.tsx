import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import * as Items from 'containers/Items';
import {
  itemDetailMockData,
  itemDetailsMockData,
} from '__mockData__/itemDetail.ts';
import Index from '../index';

describe('components/pages/AddScheduleDetail/index.tsx', () => {
  let wrapper: ShallowWrapper;

  const propsData: any = () => ({
    ...itemDetailMockData,
    navigation: {
      setParams: jest.fn(),
      navigate: jest.fn(),
    },
    route: {
      params: {
        itemId: '1',
        refresh: '',
      },
    },
  });

  beforeEach(() => {
    jest.spyOn(Items, 'useItems').mockImplementation(
      () =>
        ({
          refreshData: jest.fn(),
          itemDetails: itemDetailsMockData,
        } as any)
    );

    wrapper = shallow(<Index {...propsData()} />);
  });

  it('正常に表示されている', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
