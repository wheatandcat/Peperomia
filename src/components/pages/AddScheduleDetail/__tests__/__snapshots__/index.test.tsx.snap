// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/AddScheduleDetail/index.tsx 正常に表示されている 1`] = `
<Memo(AddScheduleDetailConnected)
  itemDetails={
    Array [
      Object {
        "kind": "park1",
        "memo": "memo1",
        "moveMinutes": 30,
        "place": "place1",
        "priority": 1,
        "title": "test1",
        "url": "url1",
      },
      Object {
        "kind": "park2",
        "memo": "memo2",
        "moveMinutes": 45,
        "place": "place2",
        "priority": 2,
        "title": "test2",
        "url": "url2",
      },
    ]
  }
  itemId="1"
  kind=""
  memo="memo"
  moveMinutes={30}
  navigation={
    Object {
      "navigate": [MockFunction],
      "setParams": [MockFunction],
    }
  }
  place="place"
  priority={1}
  refreshData={[MockFunction]}
  route={
    Object {
      "params": Object {
        "itemId": "1",
        "refresh": "",
      },
    }
  }
  showActionSheetWithOptions={[Function]}
  title="test"
  url="url"
/>
`;
